core/src/main/scala/org/apache/spark/memory/StorageMemoryPool.scala

```diff
@@ -73,27 +73,31 @@ class StorageMemoryPool(lock: Object) extends MemoryPool(lock) with Logging {
    *
    * @param blockId the ID of the block we are acquiring storage memory for
    * @param numBytesToAcquire the size of this block
-   * @param numBytesToFree the size of space to be freed through evicting blocks
+   * @param maxNumBytesToFree the maximum amount of space to be freed through evicting blocks
    * @return whether all N bytes were successfully granted.
    */
   def acquireMemory(
       blockId: BlockId,
       numBytesToAcquire: Long,
-      numBytesToFree: Long,
+      maxNumBytesToFree: Long,
       evictedBlocks: mutable.Buffer[(BlockId, BlockStatus)]): Boolean = lock.synchronized {
     assert(numBytesToAcquire >= 0)
-    assert(numBytesToFree >= 0)
+    assert(maxNumBytesToFree >= 0)
     assert(memoryUsed <= poolSize)
-    memoryStore.ensureFreeSpace(blockId, numBytesToFree, evictedBlocks)
-    // Register evicted blocks, if any, with the active task metrics
-    Option(TaskContext.get()).foreach { tc =>
-      val metrics = tc.taskMetrics()
-      val lastUpdatedBlocks = metrics.updatedBlocks.getOrElse(Seq[(BlockId, BlockStatus)]())
-      metrics.updatedBlocks = Some(lastUpdatedBlocks ++ evictedBlocks.toSeq)
+    if (numBytesToAcquire > memoryFree && maxNumBytesToFree > 0) {
+      val additionalMemoryRequired = numBytesToAcquire - memoryFree
+      memoryStore.evictBlocksToFreeSpace(
```

Comment:
after looking at the test case I think we should add a check here to do this only if there's a chance that evicting can free us enough memory, i.e.
```
val additionalMemoryRequired = math.max(0, numBytesToAcquire - memoryFree)
val numBytesToFree = math.min(maxNumBytesToFree, additionalMemoryRequired)
// Only evict blocks if there is a chance that doing so will grant us what we want
if (numBytesToFree > 0 && numBytesToFree <= memoryUsed) {
  memoryStore.evictBlocksToFreeSpace(...)
  ...
}
```
If we do this then we can't have the case where we try to free 150B from evicting blocks when there are only 100B worth of blocks.